src/state/slices/time_slice.ts

```diff
@@ -1,24 +1,32 @@
 import { StateCreator } from "zustand";
 
+import { FrameLoadResult } from "../../colorizer";
 import { decodeInt, UrlParam } from "../../colorizer/utils/url_utils";
 import { DEFAULT_PLAYBACK_FPS } from "../../constants";
 import { SerializedStoreData, SubscribableStore } from "../types";
 import { clampWithNanCheck } from "../utils/data_validation";
 import { addDerivedStateSubscriber } from "../utils/store_utils";
 import { DatasetSlice } from "./dataset_slice";
 
+import { IRenderCanvas } from "../../colorizer/IRenderCanvas";
 import TimeControls from "../../colorizer/TimeControls";
 
 export type TimeSliceState = {
-  /** The frame that is currently being loaded. If no load is happening,
+  /**
+   * The frame that is currently being loaded. If no load is happening,
    * `pendingFrame === currentFrame`.
    */
   pendingFrame: number;
   /** The currently loaded and displayed frame. */
   currentFrame: number;
   playbackFps: number;
   timeControls: TimeControls;
-  loadFrameCallback: (frame: number) => Promise<void>;
+  frameLoadCallback: IRenderCanvas["setFrame"];
```

Comment:
putting callbacks in state seems like a code smell to me. Are you doing this anywhere else? Is this ok to do?

Reply:
We're not doing this anywhere else. It comes down to frame setting needing to be reflected in state since it's accessed in so many places, but also that it needs to interact with the `canvas` to load and render frame data, which exists outside of state.

Reply:
This feels so wrong to me, but I found one zustand discussion where they say it's ok.  It just violates my philosophical sense of what "state" should be and in particular that it should be serializable.

Reply:
Something I _could_ do is put the `canvas` in a context, and have the `setFrame` take in an optional callback for the loading action. Every caller of `setFrame` would need to pull the `canvas` out of the context (or a function that wraps `setFrame` that links it to `canvas`) and pass in the canvas' `setFrame` method as a callback.

It gets a bit messier too though because TimeControls is a class that's currently in state too, and should probably be moved out of it in that case? But it needs immediate access to a lot of time-related state.